backends/p4fmt/attach.h

```diff
@@ -0,0 +1,259 @@
+    bool showIR;               /// if true dump IR as comments
+    bool withinArgument;       /// if true we are within a method call argument
+    bool noIncludes = false;   /// If true do not generate #include statements.
+                               /// Used for debugging.
+
+    struct VecPrint {
+        cstring separator;
+        cstring terminator;
+
+        VecPrint(const char *sep, const char *term) : separator(sep), terminator(term) {}
+    };
+
+    struct ListPrint {
+        cstring start;
+        cstring end;
+
+        ListPrint(const char *start, const char *end) : start(start), end(end) {}
+    };
+
+    // maintained as stacks
+    std::vector<VecPrint> vectorSeparator;
+    size_t vectorSeparator_init_apply_size = 0;
+    std::vector<ListPrint> listTerminators;
+    size_t listTerminators_init_apply_size = 0;
+
+    void setVecSep(const char *sep, const char *term = nullptr) {
+        vectorSeparator.push_back(VecPrint(sep, term));
+    }
+    void doneVec() {
+        BUG_CHECK(!vectorSeparator.empty(), "Empty vectorSeparator");
+        vectorSeparator.pop_back();
+    }
+    VecPrint getSep() {
+        BUG_CHECK(!vectorSeparator.empty(), "Empty vectorSeparator");
+        return vectorSeparator.back();
+    }
+
+    void doneList() {
+        BUG_CHECK(!listTerminators.empty(), "Empty listTerminators");
+        listTerminators.pop_back();
+    }
+    bool isSystemFile(cstring file);
+    cstring ifSystemFile(const IR::Node *node);  // return file containing node if system file
+    // dump node IR tree up to depth - in the form of a comment
+    void dump(unsigned depth, const IR::Node *node = nullptr, unsigned adjDepth = 0);
+    unsigned curDepth() const;
+
+ public:
+    // Output is constructed here
+    Util::SourceCodeBuilder &builder;
+    /* FIXME  -- simplify this by getting rid of the 'builder' object and just emitting
+     * directly to the ostream.  The SourceCodeBuilder object does not appear to add any
+     * useful functionality the ostream does not already provide; it just serves to
+     * obfuscate the code */
+    std::ostream *outStream;
+    /** If this is set to non-nullptr, some declarations
+        that come from libraries and models are not
+        emitted. */
+    cstring mainFile;
+
+    Attach(Util::SourceCodeBuilder &builder, bool showIR, cstring mainFile = nullptr)
+        : expressionPrecedence(DBPrint::Prec_Low),
+          isDeclaration(true),
+          showIR(showIR),
+          withinArgument(false),
+          builder(builder),
+          outStream(nullptr),
+          mainFile(mainFile) {
+        visitDagOnce = false;
+        setName("Attach");
+    }
+    Attach(std::ostream *outStream, bool showIR, cstring mainFile = nullptr)
+        : expressionPrecedence(DBPrint::Prec_Low),
+          isDeclaration(true),
+          showIR(showIR),
+          withinArgument(false),
+          builder(*new Util::SourceCodeBuilder()),
```

Comment:
This is really terrible. Why can't you simply *own* the source code builder? Why are you holding a non-const reference?

Reply:
This code is copied from ToP4. 
https://github.com/p4lang/p4c/blob/0adbbe73f4062a8ce5bf321637b1f0a983a9e104/frontends/p4/toP4/toP4.h#L109

A lot of the problems are already present there. The initial goal is to just contribute a modified version of that pass to allow experimentation.